File: src/main/java/de/spries/fleetcommander/model/core/universe/ShipFormation.java
package de.spries.fleetcommander.model.core.universe;

import java.util.Collection;
import java.util.stream.Collectors;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import de.spries.fleetcommander.model.core.Player;

public class ShipFormation {

	public static final int DISTANCE_PER_TURN = 8;

	private int shipCount;
	private Planet origin;
	private Planet destination;
	private Player commander;
	private int distanceTravelled = 0;
	private final double distanceOverall;

	public ShipFormation(int shipCount, Planet origin, Planet destination, Player commander) {
		if (shipCount <= 0) {
			throw new IllegalArgumentException("Must send positive number of ships");
		}
		if (origin == null || destination == null || commander == null) {
			throw new IllegalArgumentException("all parameters must be non-null");
		}
		this.shipCount = shipCount;
		this.origin = origin;
		this.destination = destination;
		this.commander = commander;
		distanceOverall = origin.distanceTo(destination);
	}

	public int getShipCount() {
		return shipCount;
	}

	public Planet getOrigin() {
		return origin;
	}

	public Planet getDestination() {
		return destination;
	}

	public Player getCommander() {
		return commander;
	}

	public boolean canJoin(ShipFormation existingFormation) {
		if (origin.equals(existingFormation.origin) && destination.equals(existingFormation.destination)
				&& commander.equals(existingFormation.commander)
				&& distanceTravelled == existingFormation.distanceTravelled) {
			return true;
		}
		return false;
	}

	public void join(ShipFormation existingFormation) {
		if (!canJoin(existingFormation)) {
			throw new IllegalArgumentException("cannot join this formation");
		}
		existingFormation.shipCount += shipCount;
		shipCount = 0;
	}

	public void travel() {
		distanceTravelled += DISTANCE_PER_TURN;
		if (hasArrived()) {
			landOnDestination();
		}
	}

	protected void landOnDestination() {
		destination.landShips(shipCount, commander);
		shipCount = 0;
	}

	public boolean hasArrived() {
		return distanceTravelled >= distanceOverall;
	}

	public int getDistanceTravelled() {
		return distanceTravelled;
	}

	@Override
	public boolean equals(Object obj) {
		return EqualsBuilder.reflectionEquals(this, obj);
	}

	@Override
	public int hashCode() {
		return HashCodeBuilder.reflectionHashCode(this);
	}

	public double getPositionX() {
		return origin.getX() + (destination.getX() - origin.getX()) * distanceTravelled / distanceOverall;
	}

	public double getPositionY() {
		return origin.getY() + (destination.getY() - origin.getY()) * distanceTravelled / distanceOverall;
	}

	public static Collection<ShipFormation> filterByCommander(Collection<ShipFormation> shipFormations, Player commander) {
		return shipFormations.parallelStream()
				.filter((s) -> s.getCommander().equals(commander)).collect(Collectors.toList());
	}

}

File: src/main/kotlin/de/spries/fleetcommander/model/core/universe/ShipFormation.kt
package de.spries.fleetcommander.model.core.universe

import de.spries.fleetcommander.model.core.Player
import de.spries.fleetcommander.model.core.common.IllegalActionException

data class ShipFormation(var shipCount: Int = 1, private val origin: Planet, private val destination: Planet, val commander: Player)
    : HasCoordinates(origin.x, origin.y) {

    private val distanceOverall: Double = origin.distanceTo(destination)
    var distanceTravelled = 0
        private set

    fun distanceRemaining() = distanceOverall - distanceTravelled

    init {
        if (shipCount <= 0) {
            throw IllegalActionException("Must send positive number of ships")
        }
    }

    fun canJoin(existingFormation: ShipFormation): Boolean {
        return (origin == existingFormation.origin
                && destination == existingFormation.destination
                && commander == existingFormation.commander
                && distanceTravelled == existingFormation.distanceTravelled)
    }

    fun join(existingFormation: ShipFormation) {
        if (!canJoin(existingFormation)) {
            throw IllegalArgumentException("cannot join this formation")
        }
        existingFormation.shipCount += shipCount
        shipCount = 0
    }

    fun travel() {
        distanceTravelled += DISTANCE_PER_TURN

        x = (origin.x + (destination.x - origin.x) * distanceTravelled / distanceOverall).toInt()
        y = (origin.y + (destination.y - origin.y) * distanceTravelled / distanceOverall).toInt()

        if (hasArrived()) {
            landOnDestination()
        }
    }

    fun landOnDestination() {
        destination.landShips(shipCount, commander)
        shipCount = 0
    }

    fun hasArrived(): Boolean {
        return distanceRemaining() <= 0
    }

    companion object {
        const val DISTANCE_PER_TURN = 8
    }
}
